feat: code generation for types based on openapi doc

andersfylling opened the pull request:
# Second draft
Second draft of code generation, see [previous version here](https://github.com/cognitedata/cognite-sdk-js/pull/740). There's a few changes such as:
 - implemented in typescript
 - unit / integration tests
 - new library (seems swagger-typescript-api is currently on "pause")
 - configuration file per service
 - a way to opt in/opt out of the package versionfile
 - ~generated types are sorted for a better git diff~ (bugged feature, wrote a something that sorts the entire json spec)
 - several smaller commands instead of one large command
 - avoids exporting already exported types (codegen per service may re-define an existing type)
 - generate types for inlined requests (with automatic naming - can be toggled off in the config file per service)

# Code (type) Generation
The SDK types should mirror the open api definition. By generating types we avoid human errors and inconsistencies in the SDK.

This also saves developers time, provides a higher likelihood of correctness as well as faster SDK updates. (This may now be automated in the future).

Whenever types generated seems incorrect, the ideal solution is that the openapi spec have to be updated - as this might indicate incorrect schema definitions. Whenever a real bug with the code generation is identified, we reproduce these thanks to the versionfile (formerly referred to as "lockfile") to reproduce the bugged behavior and come up with patches (see https://github.com/cognitedata/cognite-sdk-js/pull/740#pullrequestreview-859316537).

## Concerns

By using open api as a source, we now depend on naming conventions defined there. Changes to existing types is essentially a breaking change as the old type would no longer exist and dependents will fail to compile unless they update their imported types.

Therefore any team/service that wants this feature have to enable it manually, to become aware how this impact their work on the open api contracts.

This is a new feature and should not be assumed to be perfect. Our open API contracts also have no constraints and we don't know how certain features may affect the generated output here. We therefore have this as a opt in feature with the ability to put specific services on pause in case any temporary issues arise.

## Open API Snapshot
We require a snapshot of the open api contracts in order to generate code, this allows us to:
1. verify with github action that generated types are not manipulated / changed by hand within a pull request.
2. reprocude buggy behavior.

Each package (stable, playground, etc.) that supports code generation will have their own snapshot, shared among services. The goal is to continuously update the package snapshot, and run code generation respectively. Automatic version estimate is another problem which is not covered by this pull request - eg. should changes to types be a fix, feature or breaking change in the SDK when automating updates.

A service may specify their own snapshot if there are bugs or issues preventing it from using the package snapshot. This is done by specifying a "path" in the configuration file. This should be considered a last resort as a service is effectively put on "pause" until it's snapshot/config is manually updated.

## Config file
A configuration file is defined for the package and per service. Here a service may tweak settings to achieve desired behavior such as disabling "magic" behavior or specifying a custom open api snapshot.

## Example repo setup
Here the `documents` service uses the package open api snapshot (default), while `vision` uses their own snapshot, the `files` service have not enabled code generation as it has no configuration file (.codegen.json):
```markdown
stable/
├─ src/
│  ├─ .cognite-api-snapshot
│  ├─ types.gen.ts
│  ├─ imports.gen.ts
│  ├─ api/
│  │  ├─ vision/
│  │  │  ├─ .cognite-api-snapshot
│  │  │  ├─ .codegen.json
│  │  │  ├─ types.gen.ts
│  │  ├─ documents/
│  │  │  ├─ .codegen.json
│  │  │  ├─ types.gen.ts
│  │  ├─ files/
```

```json
{
	"service": "documents",
	"filter": {
		"serviceName": "documents"
	},
	"inlinedSchemas": {
		"autoNameRequest": true
	}
}
```

```json
{
	"service": "vision",
	"filter": {
		"serviceName": "vision"
	},
	"inlinedSchemas": {
		"autoNameRequest": false
	},
	"snapshot": {
		"path": "packages/stable/src/api/vision/.cognite-api-snapshot"
	}
}
```

The `codegen generate-types` command will generate types for the package (meaning all configured services; `documents` and `vision`). 

Example:
```console
yarn codegen generate-types --package stable
```

# General code flow
In order to generate types for a package there must be at least one service configured for code generation. Any service folder (under `./packages/{package}/api/{service}`) that contains a `.codegen.json` is considered to be a configured service (use `yarn codegen:configure` to enable code generation for your service).

Types are generated per service, with respect to their config file. The types are stored as a `types.gen.ts` file within the service folder.

A "export file" is created which explicitly exports types from each service. This is done to avoid exporting types with the same name, each service currently have their own definition for shared types.

Say you have two services `documents` and `files`. They both have defined the `LabelList` type. Since openapi schema names are globally unique, we assume that any type defined in multiple services have the same definition, so we keep track of already exported types to avoid re-exporting them.

> Fictive example
```ts
export { LabelList, FilesResponse } from './api/files/types.gen'; // types: LabelList, FilesResponse
export { DocumentsSearchResponse } from './api/documents/types.gen'; // types: LabelList, DocumentsResponse
```

# Commands
> Subject to change!

```markdown
# yarn codegen 
 * generate-types: generates types for all configured services
 * fetch-latest: fetches the latest open api and saves it as a snapshot
 * configure: generates a default config for the service or package
```

# Scenarios
> Using the "documents" service for these examples

## Enable code generation for a package
```console
# configure the package to use cognite open api v1
yarn codegen configure --package my-package --version v1

# fetches the latest open api contract, creating a snapshot
yarn codegen fetch-latest --package my-package

# generate types for package (generates no new types)
yarn codegen generate-types --package my-package
```

Then update the `types.ts` in the package `src` folder to export from `exports.gen.ts`:
```ts
export * from './exports.gen';
```

Update the github action script at `.github/workflows/code-gen.yaml` to run for your new package:
```yaml
package: ["stable", "playground", "my-package"]
```

## Enable code generation for a service
```console
yarn codegen configure --package stable --service documents
```

And generate types for the package and newly configured documents service:
```console
yarn codegen  generate-types --package stable
```

## Disable code generation for a service
For whatever reason the code generation have to be disabled for a service, you simply delete the `codegen.json` file for the respective service(s), and re-run code generation:
```console
rm ./packages/stable/src/api/documents/codegen.json
yarn codegen generate-types --package stable
```

## A service wants to run code generation from local openapi spec
> Here the service have already enabled code generation

This is useful when a service has just enabling code generation. Sometimes there might be issues in their openapi spec and they want to work on both their openapi contracts and their js-sdk types. The nice effect of generating types based off an openapi spec is that the developer gets a different perspective on their openapi definitions, it may be easier to identify issues when looking at code instead of a large yaml definition.

Update your service config to refer to a open api path:
```json
{
	"service": "documents",
	"filter": {
		"serviceName": "documents"
	},
	"inlinedSchemas": {
		"autoNameRequest": false
	},
	"snapshot": {
		"path": "/tmp/v1.json"
	}
}
```
> Config file is validated whenever you try to generate types

You don't have to run a snapshot command as you are pointing to a local path. The snapshot is simply read from the location when it's required, instead you can generate type right away:
```console
yarn codegen generate-types --package stable
```


# Issues

## Commands: configure & snapshot
~Currently the `yarn codegen:configure` and `yarn codegen:snapshot` is decoupled. It can be argued that the `yarn codegen:snapshot` should always read from a service configuration (when a --service argument is specified). This would help the PR review process as the config file also details where the snapshot originates from (though without a guarantee anyways).~

fixed...

## Shared types
> Types such as CogniteInternalId, LabelList, etc.

~Currently shared types are not generated by themselves. This mean that any service defined first in the export file will have ownership over a subset of the shared types definitions. If the first service uses a local outdated versionfile, this may cause the SDK to have outdated shared types, which can be a problem that needs manual intervention.~

~This can be fixed by identifying shared types and specified as the first export statement:~
 ~1. generating types for all services once~
 ~2. identify shared types~
 ~3. run type generation from the package global versionfile for identified shared types using a components/schemas filter.~
 ~4. Adjust the first line of the export file to use the shared generated types.~

~This fix may also cause services with oudated shared types, to now have incorrect types. However, I'd argue that we should prefer our shared types to be as up to date as possible and instead push relevant service to stay up to date.~
_Fixed in commit https://github.com/cognitedata/cognite-sdk-js/pull/801/commits/bf96dd94a58c7138cd2bcc6c2f68bc552d4ff9f8_


# Other

Required for this PR to pass:
 - Disable repository/branch requirement for node v12 tests
	A new workflow job called `tests-legacy` was added instead. Code generation doesn't work with node v12, and we won't add support for generating types via node v12 anyways. 

related tasks: SBS-4943, SBS-4945, SBS-4950, SBS-4947, SBS-5107, SBS-5122

andersfylling committed:
docs: add documentation

andersfylling committed:
chore: add dependencies

andersfylling committed:
chore: add codegen package

andersfylling committed:
chore: configure ts-node

andersfylling committed:
chore: yarn commands/scripts using yargs

andersfylling committed:
chore: add codegen logic

andersfylling committed:
chore: remove build command

andersfylling committed:
chore: add nodejs version guard

andersfylling committed:
test: use node 14, 16 and 18

andersfylling committed:
test: drop v18

andersfylling committed:
test: cyclic references

andersfylling committed:
chore: setup documents service

andersfylling committed:
chore: minor fixes

andersfylling committed:
docs: cleanup

andersfylling committed:
test: github action to verify generated files

andersfylling committed:
chore: generated shared types for package

henrist commented on .github/workflows/code-gen.yaml:
```diff
@@ -0,0 +1,41 @@
+          node-version: 16.x
+
+      - name: Install Dependencies
+        run: yarn install
```
```suggestion
        run: yarn install --frozen-lockfile
```

henrist replied:
Can we also look at the existing workflow files and make them more similar?

henrist commented on .github/workflows/code-gen.yaml:
```diff
@@ -0,0 +1,41 @@
+        run: yarn codegen --package=${{ matrix.package }} generate
+      
+      - name: Fix lint issues
+        run: yarn lint:fix --since=master
```
I don't see why linting is mixed in here? I'd not expect it to be relevant (at least as a separate command).

I suggest to exclude generated files from linting (e.g. in `.eslintignore`) instead of running the linter on the files. Otherwise I think this should be done as part of the generate command, and not require another lint run (I think I prefer to exclude it).

andersfylling replied:
generating that export file. I don't really use templates, so it seems easier to just run prettier as it automatically fixes this.

henrist replied:
Can we just accept the output the code generation produces?

andersfylling replied:
IT has to be formatted at some point. I'm not sure how to execute prettier on the string before saving it, so this seemed easier. `yarn lint...` has to be run anyways for a PR to be approved.

henrist replied:
Why does it have to be formatted? You can just place the file in .eslintignore

andersfylling replied:
I like that idea of disabling linting for generated files in general, however that means this wont add the copyright comment at the top of the file anymore (I figure).

henrist replied:
If you'd like to preserve that I'm sure it's easy to just add that line manually, rather than having to call into the code formatter. I don't like that you have to run two commands when working on this, as you won't get the instant feedback of what changes when you regenerate code, and the (lint fix) command might change other files you are working on.

andersfylling replied:
Sure, it's probably not a big deal that the copyright statement is outdated after a year or more, so ill update it.

henrist commented on .github/workflows/tests.yaml:
```diff
@@ -156,7 +156,7 @@ jobs:
     strategy:
       matrix:
-        node-version: [12, 14]
+        node-version: [14, 16]
```
Let's mark the PR for breaking change?

henrist commented on package.json:
```diff
@@ -67,11 +72,16 @@
+    "openapi-types": "^11.0.1",
     "prettier": "^2.5.1",
     "rollup": "^2.22.2",
+    "swagger-typescript-api-nextgen": "^10.0.1",
```
I'm a bit reluctant into relying on newly forks. Is `swagger-typescript-api` the right tool for the job?

andersfylling replied:
The author has been inactive since February. I'm uncertain why. There were a few features that weren't published yet, so I made the switch. I wanted sort types (broken) and auto loading the prettier config.

henrist replied:
I see. I'm a bit afraid that we will run into the same problem with the fork. I think if really needed we could always fork this out outself (should be fine since this is a dev dep), however I'm asking the question if we are choosing the right tool or if there are other more well-maintained alternatives that would suit us better.

andersfylling replied:
That is a good point, I'm leaving this unresolved to make others aware.

We talked about it, but I'll re-iterate for other reviewers:
All types that are generated have been moved into the `components.schemas` of the open api definition before running code generation. This should allow us to move to different libraries without issues. Ofcourse the way they may generate code (like prefering types for enums, or type comments may differ), but the future will tell if this becomes an issue.

henrist commented on packages/codegen/package.json:
```diff
@@ -0,0 +1,26 @@
+    "test:debug": "node --inspect-brk ../../node_modules/.bin/jest --config=../../jest.config.js --testPathPattern=/codegen/ --runInBand",
+    "lint": "eslint 'src/**/*.ts'",
+    "lint:fix": "yarn lint --fix",
+    "watch": "rollup -cw",
```
clean, rollup, es-check and docs doesn't seem relevant for this package

henrist commented on packages/codegen/package.json:
```diff
@@ -0,0 +1,26 @@
+  },
+  "files": [
+    "dist"
+  ]
```
Probably not relevant as we don't pack (at least shouldn't need to)

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,342 @@
+    await file.cleanup();
+  };
+
+  // private operationResponseWithoutRef = (operation: OpenApiOperationObject): boolean => {
```
Cleanup?

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,342 @@
+  private autoNameInlinedResponses = (): [string, OpenApiSchema][] => {
+    return [];
+    // console.error('inline responses are not currently supported');
+    // process.exit(1);
```
Cleanup?

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,342 @@
+    //  8. run swagger code gen on the filtered open api spec
+    //  9. use swagger's onCreateComponent hook to remove types to be imported from core
+    // 10. update file comment and imports
+    // 11. write to file
```
Cleanup?

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,342 @@
+// Copyright 2020 Cognite AS
```
2022?

andersfylling replied:
Seems its a general issue in the repo, it's automatically done by eslint. see .eslintrc.js, I'm leaving it as is as it's a seperate issue.

andersfylling replied:
Ay, I just noticed that the .eslintrc.js doesn't update it if there is already a year specified. Updating comments.

henrist commented on packages/codegen/src/commands.ts:
```diff
@@ -0,0 +1,269 @@
+    try {
+      await config.delete();
+    } catch (error) {
+      throw new Error('Config file does not exist - did nothing');
```
This might hide other errors, e.g. a read only file system. I suggest checking if the file exists explicitly rather than assuming all errors are not founds.

henrist commented on packages/codegen/src/configuration.ts:
```diff
@@ -0,0 +1,64 @@
+    config: ConfigurationOptions
+  ): Promise<ConfigurationOptions> => {
+    this.validate(config);
+    const json = JSON.stringify(config, null, 4);
```
We usually use 2 spaces. Se e.g. `package.json` files and `.editorconfig`

henrist commented on packages/codegen/src/configuration.ts:
```diff
@@ -0,0 +1,64 @@
+    this.validate(config);
+    const json = JSON.stringify(config, null, 4);
+
+    await fs.writeFile(this.path, json, { flag: 'wx' });
```
Do we care if the file already exists (the `x` flag)? I guess no flags are needed here

henrist replied:
It would be nice to add training newlines as well, according to the rule in `.editorconfig`. Especially since this file is legit to edit manually.

henrist commented on packages/codegen/typedoc.js:
```diff
@@ -0,0 +1,17 @@
+module.exports = {
```
Is typedoc needed for this package? I think we can remove

henrist commented on packages/codegen/src/versionfile.ts:
```diff
@@ -0,0 +1,82 @@
+    return `https://storage.googleapis.com/cognitedata-api-docs/dist/${this.options.version}.json`;
+  };
+
+  public downloadFromURL = async (url?: string): Promise<OpenApiDocument> => {
```
nit: URL vs Url (I prefer Url)

henrist commented on packages/codegen/src/versionfile.ts:
```diff
@@ -0,0 +1,82 @@
+ * To keep the version file "small", the irrelvant paths are filtered out before saving.
+ */
+export class VersionFileManager {
+  public static readonly filename = '.cognite-api-version';
```
I find this name a bit confusing, especially since the "version" field inside the file never changes. And maybe the file shouldn't be hidden by default?

In my head what we are looking at is a snapshot from some time. How about something like `openapi-spec-snapshot.json` ?

andersfylling replied:
Yeah, snapshot is a better fit. Updated

henrist commented on packages/codegen/src/versionfile.ts:
```diff
@@ -0,0 +1,82 @@
+export type LocalVersionFileOptions = {
+  path: string;
+  filename?: string;
+};
```
nit: I'd call for using interfaces instead of type aliases thoughout, unless you need the power of expressions. See recommendations in https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases

henrist commented on .github/workflows/code-gen.yaml:
```diff
@@ -0,0 +1,41 @@
+  stable:
+    strategy:
+      matrix:
+        package: ["stable"]
```
I suggest we include playground as well, so that's not missed later.

Do we need to have a build matrix? I'd think that we could have one command generate types for everything (or delegate if needed) - that would mean we remove the package argument. Maybe each package.json file that needs codegen should have a run-script for it, and use Lerna to delegate?

andersfylling replied:
Is the package argument then implicit via lerna? => Would that cause a command to affect all packages at the same time?

henrist replied:
It would cause the package argument to go away, since you will run the codegen on one package at a time

henrist replied:
Probably just finish what's began now instead of doing too much changes to this.

andersfylling replied:
Just out of curiosity.

so basically `cd ./packages/stable && yarn codegen ...`? 

And then `yarn codegen generate` triggers on all packages?

henrist replied:
See how the other commands like build and lint works

henrist commented on CONTRIBUTING.md:
````diff
@@ -55,6 +55,41 @@ docs(contributing-readme): add example of commit with subject line
+A versionfile has two scopes; local (service specific) or global (package specific). Services that links to a package global versionfile are all updated at the same time. If you require to update types for your service only, independently from other services in the package, it's recommended to use a local versionfile.
+```sh
+# configure your service to have a local versionfile
+> yarn codegen:configure --package=stable --service=documents --versionfile-scope=local update
````
Since this is supposed to be a script (`sh` above) the `> `s should be removed.

andersfylling replied:
I just discovered there's a `console` option. Adjusted.

henrist commented on scripts/codegen/configure.ts:
```diff
@@ -0,0 +1,45 @@
+        if (typeof argv['versionfile-scope'] !== "undefined") {
+            config["scope"] = argv['versionfile-scope'];
+        }
+        await cmd.update(config);
```
This will reinitialize the file on every update. I suggest we patch the file instead.

henrist commented on scripts/codegen/configure.ts:
```diff
@@ -0,0 +1,45 @@
+    }
+}).command("update", "update a configuration", (y) => y, async (argv) => {
+    try {
+        let config = argv;
```
This will include lots of other values that is present in `argv`. I think it's better to select what you expect and build a new object.

henrist commented on scripts/codegen/codegen.ts:
```diff
@@ -0,0 +1,28 @@
+
+    try {
+        await cmd.generate({...argv});
+        console.log("ok");
```
remove?

henrist commented on scripts/codegen/notes.txt:
```diff
@@ -0,0 +1,45 @@
+# generate a new versionfile in the stable package. This file points to openapi v1 and has created a configuration for the documents service.
```
Remove file?

henrist commented on tsconfig.common.json:
```diff
@@ -1,4 +1,19 @@
+  "ts-node": {
+    //"extends": "ts-node/node16/tsconfig.json",
+    // It is faster to skip typechecking.
+    // Remove if you want ts-node to do typechecking.
```
Let's remove all these comments?

henrist commented on tsconfig.common.json:
```diff
@@ -1,4 +1,19 @@
+    // It is faster to skip typechecking.
+    // Remove if you want ts-node to do typechecking.
+    "transpileOnly": true,
+    "files": true,
```
nit: I don't think this is needed?

henrist commented on scripts/codegen/codegen.ts:
```diff
@@ -0,0 +1,28 @@
+import { CodeGenCommand } from '../../packages/codegen/src';
```
Just an idea: I think it would be nice to move all this codegen scripts into a single cli entrypoint on the codegen package

andersfylling replied:
I tried using the CommandModule as you suggested. However, I feel more confused now. Especially when running the help argument for a sub command. So I put this change on ice, perhaps we can take a look together at a later point as I'm finding the behavior a bit weird right now?

henrist replied:
What is confusing? Sure, we can look at it later?

henrist commented on packages/stable/src/exports.gen.ts:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Cognite AS
+export {} from './types.gen';
```
Something odd here

andersfylling replied:
Ah. It's because code has only been generated for only one service. This gets populated as soon as we have codegen for an additional service, so I'm not sure it's worth the extra effort of making generated code nice when only one service has it enabled.

henrist commented on packages/stable/src/exports.gen.ts:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Cognite AS
```
Should we add a `linguist-generated=true` rule in `.gitattributes` to match on these autogenerated files? Similar as how baker files etc are done in other repos

henrist commented on packages/codegen/src/commands.ts:
```diff
@@ -0,0 +1,338 @@
+
+    try {
+      await config.delete();
+      await config.write(this.defaultConfig(options));
```
I suggest overwriting the file instead. This helps certain editors keep the correct (inode) reference to the file

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+  private runSwaggerGenerator = async (): Promise<void> => {
+    const fileComment = `// Do not modify this file!
+// It was generated by the command "yarn codegen".
+// Instead update the code generation logic or the open api constracts.\n\n`;
```
```suggestion
// Instead update the code generation logic or the open api contracts.\n\n`;
```

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+import { ConfigurationOptions } from './configuration';
+
+/**
+ * PathFilter creates a new set of openapi paths that pass the filter implementation.
```
This doesn't look to belong here

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+  static outputFileName = 'types.gen.ts';
+  readonly versionFile: OpenApiDocument;
+
+  constructor(versionFile: OpenApiDocument, readonly options: CodeGenOptions) {
```
`options` unused?

henrist replied:
Forget my comment - I didn't know we could declare class members with `readonly` in a constructor parameter

andersfylling replied:
it's a bit of sugar, might be cleaner to just drop it tbf

henrist replied:
It's fine. I just weren't aware of the feature.

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+export const PassThroughFilter: StringFilter = (): boolean => true;
+
+export const ServiceNameFilter = (service: string): StringFilter => {
+  const r = new RegExp('^/api/.+/projects/{project}/' + service);
```
This actually matches for "service name prefix". May I suggest we append `($|/)` to make it match on equals instead?

andersfylling replied:
isn't `$` for end of string identification? perhaps `\/?` instead.

henrist replied:
That wont change anything since you dont match until the end of the string, so it will match the same as before

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+
+export class CodeGen {
+  static outputFileName = 'types.gen.ts';
+  readonly versionFile: OpenApiDocument;
```
I think in this context it would make more sense to just name this "specFile" or something. I find the "version" term more confusing than helpful

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+  constructor(versionFile: OpenApiDocument, readonly options: CodeGenOptions) {
+    this.versionFile = JSON.parse(
+      JSON.stringify(versionFile)
+    ) as OpenApiDocument;
```
Why is this serialized and deserialized? It seems to be the same types input and output

henrist replied:
If your intent is to deep copy maybe you can use structuredClone instead: https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

henrist commented on scripts/codegen/versionfile.ts:
```diff
@@ -0,0 +1,37 @@
+    package: { type: 'string', demand: true,  describe: "Js sdk package (stable, beta, etc.)" },
+    service: { type: 'string', demand: false, describe: "REST service to generate types for" },
+    version: { type: 'string', demand: false, describe: "Cognite api version (v1, playground, etc.)" },
+    "from-path": { type: 'string', demand: false, describe: "Specify a local openapi spec instead" },
```
This is nice so it's easy to work with the spec and see what changes are caused by it!

henrist commented on scripts/codegen/configure.ts:
```diff
@@ -0,0 +1,45 @@
+}).command("delete", "delete a configuration", (y) => y, async (argv) => {
+    try {
+        await cmd.delete({...argv});
+        console.log("ok");
```
Remove?

henrist replied:
See other places too

henrist commented on packages/stable/src/api/documents/types.gen.ts:
```diff
@@ -48,120 +43,20 @@ export type DocumentListRequest = DocumentListFilter &
   DocumentListLimit &
   DocumentCursor;
 
-export interface DocumentSearch {
```
Just an open question: Is it possible (easy) to sort the types by name or something, to avoid types moving around?

andersfylling replied:
Yeah, the sorting option for the codegen library is buggy. I figure I can just sort the entire openapi object recursively, just need to check performance first..

henrist replied:
Let's scope it out for now. If we see frequent changes to the ordering and that's annoying we can look into it later.

henrist commented on CONTRIBUTING.md:
```diff
@@ -55,6 +55,41 @@ docs(contributing-readme): add example of commit with subject line
+The scripts prefix `codegen` are reserved for type generation. The core logic is found in `packages/codegen` with a deeper explenation.
+
+To quickly setup type generation for your service you may run the generate command with `--auto`:
+> Note: using `--auto` will setup a local versionfile and not the package global versionfile. You can later switch to the global versionfile once you have confirmed type generation works correctly for your service (you may also have to update your openapi spec).
```
I suggest we swap this - always prefer global file but opt in to local file as a strategy for conflict resolution.

andersfylling replied:
Wasn't that my suggestion earlier today - or am I misunderstanding something?

henrist replied:
Might be

andersfylling replied:
From what I can tell, it is. Perhaps you were thinking of something additional to my suggestion earlier? I've updated the PR description to reflect my suggestion so please take a look when you get back from vacay :sun_with_face:

andersfylling replied:
From what I can see, it is. Perhaps you were thinking about something additional?

I've updated the PR description and I'm in the progress of updating the code. Please take a look when you're back from vacay :sun_with_face:

henrist replied:
I see this content is no longer present. It seems now that you have to change the config to be able to regenerate with a custom path, e.g. for testing some change? I liked the way it was easy previously to generate types for local spec with least effort (e.g. while working on service contracts) - and I'm not sure if it needs to be coupled to a config as it would only be temporary?

henrist commented on packages/codegen/README.md:
```diff
@@ -0,0 +1,28 @@
+To run code generation for a package the script behaves as follows:
+
+ 1. Identify any services with a config file.
+ 2. Evaluate the relevant lockfile for identified services.
```
rephrase lockfile

henrist commented on packages/codegen/README.md:
```diff
@@ -0,0 +1,28 @@
+ 1. Identify any services with a config file.
+ 2. Evaluate the relevant lockfile for identified services.
+ 3. Generate relevant types/interfaces.
+ 4. Write types to a types.gen.ts file.
```
In general - can we add ``` ` ``` around filenames like this?

henrist commented on packages/codegen/README.md:
```diff
@@ -0,0 +1,28 @@
+ 5. Create an export statement for each service, that avoids exporting existing types.
+ 6. Write export statements to a types.gen.ts file at the package level.
+
+See codegen.ts#main for high level understanding of the flow.
```
This reference looks incorrect (and I guess it's a bit redundant)

henrist commented on packages/codegen/rollup.config.js:
```diff
@@ -0,0 +1,44 @@
+import typescript from '@rollup/plugin-typescript';
```
We don't use the output of rollup for this package so maybe just remove?

henrist commented on packages/codegen/src/configuration.ts:
```diff
@@ -0,0 +1,64 @@
+  version: string;
+  service: string;
+  autoNameInlinedRequest: boolean;
+  scope: string;
```
```suggestion
  scope: 'local' | 'global';
```

And maybe consider changing to `service` and `package` as scopes? I find `local` and `global` a bit confusing

andersfylling replied:
I like this idea

henrist commented on packages/codegen/src/commands.ts:
```diff
@@ -0,0 +1,338 @@
+      await this.generateForSingleService({
+        ...options,
+        service: options.service,
+      });
```
This doesn't have any effect?

henrist replied:
I'm not sure if we should even support generating for only one service. It adds complexity to figuring out shared types (which I think is incorrect now if we were to only generate for one service).

andersfylling replied:
> which I think is incorrect now if we were to only generate for one service

I don't think we will be generating types for only one service in the future anyways.

henrist replied:
Any reason this is still left around? The code still seems wrong/incomplete around shared types

henrist commented on packages/codegen/src/commands.ts:
```diff
@@ -0,0 +1,338 @@
+    fromDirectory: string,
+    types: string[],
+    blacklist: string[]
+  ): [string, string[]] => {
```
I'd recommend returning an object rather than array. It helps a lot to understand what data there is without look at the implementation.

Something like:

```ts
private createExportStatement = (
    fromDirectory: string,
    types: string[],
    blacklist: string[]
  ): { code: string, skipped: string[] } => {
    // ...
  }

// and then
const { code } = createExportStatement(...)
```

henrist commented on CONTRIBUTING.md:
````diff
@@ -55,6 +55,41 @@ docs(contributing-readme): add example of commit with subject line
+yarn codegen --package=stable --version=v1 --service=documents --auto generate
+```
+
+A service must have a versionfile (essentially just a copy of the openapi spec) and a configuration (which openapi paths to parse). After
````
I think it's a bit "hidden" in the text _which_ endpoints types are actually generated for. Can we make that more clear? Ideally I think we could shorten down all this text around.

Can a "service" start by only generating types for a single endpoint, or is it "all or nothing"?

andersfylling replied:
A service has to, currently, generate types for all service owned endpoints.

I made the config/StringFilter option in mind so we could extend this if we wanted to.

henrist reviewed (commented):
This is a lot of work hence quite a few comments. I haven't finished looking through the "code generation" code yet. I'd like to point out that this brings a lot of complexity (and 1700+ lines of code, excluding fixtures/generated), and I wonder if we could have solved it a bit simpler with less complexity initially. I want to challenge you to see if there is some code/text that can be stripped down/simplified without loosing essential functionality.

The CLI in general looks nice. Maybe we can have it all through one entrypoint, so `yarn codegen` could hold the important documentation and access to all commands?

I think the feedback is coming a bit late here. It would be nice if we could arrange feedback earlier in the process or maybe do some pair programming or design talk earlier to drive some discussions.

andersfylling commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,352 @@
+  return (path: string): boolean => r.test(path);
+};
+
+export type CodeGenOptions = {
```
I think this could be simplified.

andersfylling committed:
chore: cleanup after review comments

andersfylling committed:
chore: update packages/codegen/src/codegen.ts

Co-authored-by: Henrik Steen <henrist@henrist.net>

andersfylling committed:
test: fix expected output ts file

andersfylling committed:
test: fix expected output ts file

andersfylling committed:
test: update .github/workflows/code-gen.yaml

Co-authored-by: Henrik Steen <henrist@henrist.net>

cognite-bulldozer[bot] committed:
Merge master into codegen-v2

cognite-bulldozer[bot] committed:
Merge refs/heads/master into codegen-v2

cognite-bulldozer[bot] committed:
Merge refs/heads/master into codegen-v2

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,316 @@
+    this.filterPaths();
+    this.removeSchemasWithoutPathRelations();
+    this.copyResponsesToSchemas();
+    this.copyQueryParametersToSchemas();
```
I'm not a large fan of code like this. It makes the flow of operations and effects harder to follow, and e.g. breaks in the case of method being called twice and such, since it's building state in the object.

I'd recommend being explicit about the values you pass around, as in removing `this.versionFile` at least. I'm neutral on doing mutations on the values in the methods though - that should be fine as long as we are working on a deep copy.

henrist replied:
As an example here it was tricky to understand that the `this.versionFile` used in `runSwaggerGenerator` is _not_ the same as what the object was initialized with. I'd say that in the context of `runSwaggerGenerator` that variables has a bad and confusing name. It could be a lot more clear if the signature was something like `private runSwaggerGenerator = async (transformedSpec: OpenApiDocument) =>` instead

I also think it would be more clear if the side effects (such as writing `types.gen.ts`) was not hidden so deep and mixed with other operations. It's also confusing in regard to `generateTypesFromSchemas` vs `generateTypes`. Invoking only the first will not do what the second is doing, but both end up writing a types file (in the same place with different content?).

andersfylling replied:
`generateTypesFromSchemas` will only look at the schemas. While `generateTypes` will try to derive more types than just those defined in `schemas`.

andersfylling replied:
I don't think schemas are altered in `generateTypes`, only new schemas are added (unless unused ones are filtered out).

henrist commented on packages/codegen/src/__tests__/testCodeGen.uint.spec.ts:
```diff
@@ -0,0 +1,174 @@
+// import path from 'path';
```
Having `test` in the name is a bit redundant as it's `.spec.` that matters. So I suggest renaming these files to just `codeGen.uint.spec.ts` etc)

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,316 @@
+
+export const PassThroughFilter: StringFilter = (): boolean => true;
+
+export const ServiceNameFilter = (service: string): StringFilter => {
```
nit: These are just functions but looks like classes or types with their capital first letter - `passThroughFilter` and `serviceNameFilter`

henrist commented on packages/codegen/src/codegen.ts:
```diff
@@ -0,0 +1,316 @@
+        console.error(
+          `unable to use automatic name for inlined request type. "${typeName}" already exists in the component schemas`
+        );
+        process.exit(1);
```
I think throwing an exception is better here - since you're not really in the CLI level here

henrist commented on packages/codegen/src/commands.ts:
```diff
@@ -0,0 +1,338 @@
+    };
+
+    const document = await download();
+    await vfm.write(document);
```
nit: If you can I'd recommend using ternary operator over control flow - it often simplifies things (and notice how to check for undefined):

```suggestion
    const document =
      options['from-path'] !== undefined
        ? await vfm.downloadFromPath({ path: options['from-path'] })
        : await vfm.downloadFromURL();

    await vfm.write(document);
```

henrist commented on packages/codegen/src/openapi.ts:
```diff
@@ -0,0 +1,125 @@
+  return Object.values(filtered)
+    .map(operationsInPath)
+    .reduce((acc, v) => acc.concat(v), []);
+}
```
These seem unused?

andersfylling replied:
See codegen.ts line 88

henrist replied:
Still unused. Are you mixin this with the next method `operationsInPath` ?

andersfylling committed:
chore: minor improvements

andersfylling committed:
chore: minor improvements

andersfylling committed:
style: fix

andersfylling committed:
chore: run code gen + fixes

andersfylling committed:
chore: wip

andersfylling committed:
chore: simpler commands + always read from config

andersfylling committed:
chore: enable codegen for playground & playground/documents service

andersfylling committed:
fix: capitalize first char of schemas

andersfylling committed:
chore: sort json spec

andersfylling committed:
chore: fix potential null issue + regenerate types

andersfylling committed:
test: fix sorted values

andersfylling committed:
test: add --detectOpenHandles

andersfylling committed:
chore: remove dead args

andersfylling committed:
chore: ignore exports.gen.ts

henrist commented on .github/workflows/code-gen.yaml:
```diff
@@ -0,0 +1,41 @@
+name: Code Generation
```
nit: let's rename this file to `codegen.yaml` so we don't have both `codegen` and `code-gen` ?

henrist commented on scripts/codegen/codegen.ts:
```diff
@@ -0,0 +1,22 @@
+const cmd = new CodeGenCommand();
+
+const parser = yargsBase().options({
+    package: { type: 'string', demand: true,  describe: "Js sdk package (stable, beta, etc.)" },
```
These files (in the top level `scripts` dir) is not linted or formatted. I think the easiest fix is to move it into the codegen package so it will be linted by its setup

andersfylling replied:
I assume you're handling this now ^^,

andersfylling committed:
chore: remove openHandle

andersfylling committed:
chore: simply typeof undefined check

andersfylling committed:
chore: use null check instead of undefined

andersfylling committed:
chore: title case error messages

andersfylling committed:
chore: comments

andersfylling committed:
chore: cleanup variable names

andersfylling committed:
chore: rename SnapshotOnline

andersfylling committed:
chore: general cleanup + use interface instead of type

andersfylling committed:
chore: remove dead subcommands

andersfylling committed:
chore: rename code-gen.yaml to codegen.yaml

andersfylling committed:
chore: rename code-gen.yaml to codegen.yaml

andersfylling committed:
chore: move snapshot command into snapshot.ts

andersfylling committed:
chore: move config command into configuration.ts

andersfylling committed:
chore: rename commands.ts to generate.ts

andersfylling committed:
chore: rename commands.ts to generate.ts

cognite-bulldozer[bot] committed:
Merge refs/heads/master into codegen-v2

cognite-bulldozer[bot] committed:
Merge refs/heads/master into codegen-v2

andersfylling committed:
chore: cleanup + remove need to cast config types

andersfylling committed:
chore: update snapshots and rerun code generation

andersfylling committed:
chore: add eslintignore for playground pkg

henrist commented on packages/playground/.eslintignore:
```diff
@@ -0,0 +1 @@
+packages/*/src/exports.gen.ts
```
Can you check if you have EditorConfig set up? This should have a trailing newline

henrist committed:
chore: codegen improvements (#820)

* chore: rework and simplify CLI for codegen

* chore: drop the command classes in favor of top level functions

* chore: rewrite docs

* chore: remove obsolete esCheck run-script

* chore: remove obsolete tsconfig.build.json file

* chore: rename snapshot file and set linguist-generated

* chore: pretty format openapi snapshot

This makes it easier to understand what has actually been
updated in the snapshot.

* chore: minor typo

* chore: rephrase "open api"

* chore: remove node version check but add engines

* chore: set version in codegen package

To avoid yarn complaining.

* chore: re-hide the snapshot file

* chore: fix configure description

andersfylling committed:
style: fix

andersfylling committed:
chore: fix workflow file

andersfylling committed:
Merge branch 'master' into codegen-v2

andersfylling committed:
test: don't use node v12 on the codegen package

andersfylling committed:
chore: move engines directive to codegen

cognite-ornellas committed:
Merge branch 'master' into codegen-v2

andersfylling committed:
chore: update

andersfylling committed:
chore: update readme
